File: LeetCode-Java/Valid Parentheses.java
public class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> record = new HashMap<Character, Character>();
        record.put(')', '(');
        record.put('}', '{');
        record.put(']', '[');
        Stack<Character> charStack = new Stack<Character>();
        for(int i = 0;i < s.length();i ++)
        {
            char c = s.charAt(i);
            if((!charStack.empty()) && charStack.peek() == record.get(c))
            {
                charStack.pop();
            }
            else
            {
                charStack.push(c);
            }
        }
        return charStack.empty();
    }
}

File: Leetcode-cpp/valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> record;
        unordered_map<char, char> ref;
        ref[')'] = '(';
        ref['}'] = '{';
        ref[']'] = '[';
        for (auto val : s) {
            if (ref.find(val) == ref.end()) {
                record.push(val);
            } else {
                if (record.empty() || ref[val] != record.top()) {
                    return false;
                } else {
                    record.pop();
                }
            }
        }
        return record.empty();
    }
};